File: connectors/notification-hub/app/novu_client.py
"""Relay to the Novu API via a direct REST call rather than a Novu SDK,
to avoid a heavy dependency in this thin connector (study 2.1 line 382:
"Novu for the notification center UI ... fed by custom connectors").
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any, Optional

import httpx

from .types import NormalizedEvent

NOVU_API_URL = os.environ.get("NOVU_API_URL", "https://api.novu.co/v1")
NOVU_API_KEY = os.environ.get("NOVU_API_KEY", "")
# Workflow trigger identifier configured on the Novu side, shared by all sources.
NOVU_WORKFLOW_ID = os.environ.get("NOVU_WORKFLOW_ID", "libre365-unified-notification")


@dataclass
class NovuTriggerResult:
    ok: bool
    status: int
    body: Optional[Any] = None


async def trigger_novu_notification(
    event: NormalizedEvent,
    client: Optional[httpx.AsyncClient] = None,
) -> NovuTriggerResult:
    payload = {
        "name": NOVU_WORKFLOW_ID,
        "to": {"subscriberId": event["userId"]},
        "payload": {
            "source": event["source"],
            "eventType": event["eventType"],
            "title": event["title"],
            "body": event["body"],
            "actionUrl": event["actionUrl"],
            "timestamp": event["timestamp"],
        },
    }
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"ApiKey {NOVU_API_KEY}",
    }

    if client is not None:
        response = await client.post(
            f"{NOVU_API_URL}/events/trigger", json=payload, headers=headers
        )
    else:
        async with httpx.AsyncClient() as owned_client:
            response = await owned_client.post(
                f"{NOVU_API_URL}/events/trigger", json=payload, headers=headers
            )

    try:
        body = response.json()
    except ValueError:
        body = None

    return NovuTriggerResult(ok=response.is_success, status=response.status_code, body=body)

File: connectors/notification-hub/src/novu-client.ts
import { NormalizedEvent } from "./types";

/**
 * Relais vers l'API Novu via appel REST direct plutot que le SDK @novu/node,
 * pour eviter une dependance lourde dans ce connecteur mince (etude 2.1 ligne 382 :
 * "Novu pour l'UI de centre de notif ... alimente par des connecteurs custom").
 */
const NOVU_API_URL = process.env.NOVU_API_URL ?? "https://api.novu.co/v1";
const NOVU_API_KEY = process.env.NOVU_API_KEY ?? "";
// Workflow trigger identifier configure cote Novu, commun a toutes les sources.
const NOVU_WORKFLOW_ID = process.env.NOVU_WORKFLOW_ID ?? "open365-unified-notification";

export interface NovuTriggerResult {
  ok: boolean;
  status: number;
  body?: unknown;
}

export async function triggerNovuNotification(
  event: NormalizedEvent,
  fetchImpl: typeof fetch = fetch
): Promise<NovuTriggerResult> {
  const response = await fetchImpl(`${NOVU_API_URL}/events/trigger`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `ApiKey ${NOVU_API_KEY}`,
    },
    body: JSON.stringify({
      name: NOVU_WORKFLOW_ID,
      to: { subscriberId: event.userId },
      payload: {
        source: event.source,
        eventType: event.eventType,
        title: event.title,
        body: event.body,
        actionUrl: event.actionUrl,
        timestamp: event.timestamp,
      },
    }),
  });

  let body: unknown;
  try {
    body = await response.json();
  } catch {
    body = undefined;
  }

  return { ok: response.ok, status: response.status, body };
}
